src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs

```diff
@@ -1893,4 +1911,20 @@ internal void OnBeforeQueryInvoke(BeforeQueryEventArgs args)
             return (indexName, collectionName);
         }
     }
+
+    internal struct AsyncTaskHolder : IDisposable
+    {
+        private readonly InMemoryDocumentSessionOperations _session;
+
+        public AsyncTaskHolder(InMemoryDocumentSessionOperations session)
+        {
+            _session = session;
+            Interlocked.Increment(ref _session._asyncTasksCounter);
```

Comment:
Here, we also want to error if we have a non zero old value.
We don't allow concurrent async tasks in the async session.
Only a single such task is allowed.

Reply:
done